x/stakeibc/keeper/unbonding.go

```diff
@@ -309,6 +309,61 @@ func (k Keeper) GetUnbondingICAMessages(
 	return msgs, unbondings, nil
 }
 
+// Submit undelegate ICA messages in small batches to reduce the gas size per tx
+func (k Keeper) BatchSubmitUndelegateICAMessages(
+	ctx sdk.Context,
+	hostZone types.HostZone,
+	epochUnbondingRecordIds []uint64,
+	msgs []proto.Message,
+	unbondings []*types.SplitUndelegation,
+	batchSize int,
+) (numTxsSubmitted uint64, err error) {
+	// Iterate the full list of messages and submit in batches
+	for start := 0; start < len(msgs); start += batchSize {
+		end := start + batchSize
+		if end > len(msgs) {
+			end = len(msgs)
+		}
+
+		msgsBatch := msgs[start:end]
+		unbondingsBatch := unbondings[start:end]
+
+		// Store the callback data
+		undelegateCallback := types.UndelegateCallback{
+			HostZoneId:              hostZone.ChainId,
+			SplitUndelegations:      unbondingsBatch,
+			EpochUnbondingRecordIds: epochUnbondingRecordIds,
+		}
+		callbackArgsBz, err := proto.Marshal(&undelegateCallback)
+		if err != nil {
+			return numTxsSubmitted, errorsmod.Wrap(err, "unable to marshal undelegate callback args")
+		}
+
+		// Submit the undelegation ICA
+		if _, err := k.SubmitTxsDayEpoch(
+			ctx,
+			hostZone.ConnectionId,
+			msgsBatch,
+			types.ICAAccountType_DELEGATION,
+			ICACallbackID_Undelegate,
+			callbackArgsBz,
+		); err != nil {
+			return numTxsSubmitted, errorsmod.Wrapf(err, "unable to submit unbonding ICA for %s", hostZone.ChainId)
+		}
+		numTxsSubmitted += 1
+
+		// flag the delegation change in progress on each validator
+		for _, unbonding := range unbondingsBatch {
+			if err := k.IncrementValidatorDelegationChangesInProgress(&hostZone, unbonding.Validator); err != nil {
```

Comment:
Thinking through this: because when building these messages we are combining the total amounts needed to be undelegated before we run the algorithm which chooses which validators to undelegate from -- it still should be the case that across the whole batch we are only sending one message to each validator at most correct?  There should not be a case where the same validator gets more than one of these unbonding messages?

Reply:
yup that's right!